langworld_db_data/adders/listed_value_adder.py

```diff
@@ -16,54 +16,98 @@ def add_listed_value(
         """
+
         if not (feature_id and new_value_en and new_value_ru):
             raise ListedValueAdderError("None of the passed strings can be empty")
 
-        id_of_new_value = self._add_to_inventory_of_listed_values(
-            feature_id=feature_id, new_value_en=new_value_en, new_value_ru=new_value_ru
-        )
-        self._mark_value_as_listed_in_feature_profiles(
-            feature_id=feature_id,
-            new_value_id=id_of_new_value,
-            new_value_ru=new_value_ru,
-            custom_values_to_rename=custom_values_to_rename,
-        )
+        try:
+            id_of_new_value = self._add_to_inventory_of_listed_values(
+                feature_id=feature_id,
+                new_value_en=new_value_en,
+                new_value_ru=new_value_ru,
+                index_to_assign=index_to_assign,
+            )
+
+            self._mark_value_as_listed_in_feature_profiles(
+                feature_id=feature_id,
+                new_value_id=id_of_new_value,
+                new_value_ru=new_value_ru,
+                custom_values_to_rename=custom_values_to_rename,
+            )
+        except ValueError as e:
+            raise ListedValueAdderError(e)
 
     def _add_to_inventory_of_listed_values(
         self,
         feature_id: str,
         new_value_en: str,
         new_value_ru: str,
+        index_to_assign: int,
     ) -> str:
+        """
+        Returns ID of new value and tuple of indices that must be incremented in feature profiles.
+        index_to_assign means number that must be assigned to the new value within the feature, ex. 13 for A-2-13.
+        If no index_to_assign is given, the new value will be added as the last one in the feature.
+        index_to_assign must be greater than 0.
+        """
+
         rows = read_dicts_from_csv(self.input_file_with_listed_values)
 
         if not [r for r in rows if r["feature_id"] == feature_id]:
             raise ListedValueAdderError(f"Feature ID {feature_id} not found")
 
-        index_of_last_row_for_given_feature = 0
-        id_of_new_value = ""
+        # Collect all indices and line numbers of given feature into value_indices_to_inventory_line_numbers
+        value_indices_to_inventory_line_numbers = (
+            self._get_indices_and_their_line_numbers_in_features_listed_values(
+                rows=rows,
+                feature_id=feature_id,
+                new_value_en=new_value_en,
+                new_value_ru=new_value_ru,
+            )
+        )
 
-        for i, row in enumerate(rows):
-            if row["feature_id"] != feature_id:
-                continue
+        # Check if index_to_assign is -1 or belongs to range of indices in value_indices_to_inventory_line_numbers
+        last_index_in_feature = value_indices_to_inventory_line_numbers[-1]["index"]
+        acceptable_indices_to_assign = [-1] + [i for i in range(1, last_index_in_feature)]
```

Comment:
```suggestion
        acceptable_indices_to_assign = [-1] + [i for i in range(1, last_index_in_feature)]
```
I don't think you have tested all edge cases here since your tests don't fail. What if user passes in the exact value of `last_index_in_feature`? It is not included in the `range(1, last_index_in_feature)`. I think the program will fail here.

Reply:
The range in list comprehension is updated so that current maximal index + 1 is now accepted by this validation. I also made a special test in `test_adders_listed_value_adder` on this case and the program passes it.